### PR TITLE
Fix #2937: SharedInformerFactory#getExistingSharedIndexInformer returns null when @Kind configured

### LINKED ISSUE #2937: Incorrect SharedInformerFactory#getExistingSharedIndexInformer
The caching key of informers has changed in the client 5.2. And as far as I can see, the lookup function is not coherent with the create functions. The function `getExistingSharedIndexInformer` is sensitive to the name I chose for the Java class of my custom resource.

For instance, if I have a custom resource under the API `com.acme/myapp:v1`, and that I have created the Java class `MyAppCustomResource` annotated to be a proper fabric8 `CustomResource`, in the following code `existingInformer` will be `null`, but I would expect that `existingInformer` is the same instance as `createdInformer`.

    SharedIndexInformer createdInformer = sharedInformerFactory.sharedIndexInformerForCustomResource(MyAppCustomResource.class, MyAppCustomResourceList.class)
    SharedIndexInformer existingInformer = sharedInformerFactory.getExistingSharedIndexInformer(MyAppCustomResource.class)

### PR DESCRIPTION
**Pull request by rohanKanojia**
Fix #2937 

Right now we're just getting plural form of Class name for
CustomResource while resolving SharedInformerFactory informer map's
keys. But this leads to issues when user can configure his kind using
`@Kind` annotation. 

This PR changes `HasMetadata` and `CustomResource` static methods so that they can be reused in different contexts and uses the newly "available" `CustomResource.getPlural` method to ensure that plural is consistently computed.

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java [hunk 1]
```diff
@@ -63,7 +62,7 @@ default String getKind() {
    * @return the computed {@code apiVersion} or {@code null} if neither {@link Group} or {@link Version} annotations are present
    * @throws IllegalArgumentException if only one of {@link Group} or {@link Version} is provided
    */
-  static String getApiVersion(Class<? extends HasMetadata> clazz) {
+  static String getApiVersion(Class<?> clazz) {
```
Comment (manusa) [inline kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java:65]: Why? In which (valid) cases is this method going to be used with a non-HasMetadata class?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/CustomResource.java [hunk 2]
```diff
@@ -180,7 +180,7 @@ public void setMetadata(ObjectMeta metadata) {
    * @param clazz the CustomResource whose plural form we want to retrieve
    * @return the plural form defined by the {@link Plural} annotation or a computed default value
    */
-  public static String getPlural(Class<? extends HasMetadata> clazz) {
+  public static String getPlural(Class<?> clazz) {
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/CustomResource.java:183]: Same as https://github.com/fabric8io/kubernetes-client/pull/2952/files#r607517131

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java [hunk 3]
```diff
@@ -347,7 +346,12 @@ private static String getKeyStrForField(String str) {
     return keyBuilder.toString();
   }
 
-  private <T extends HasMetadata, L extends KubernetesResourceList<T>> BaseOperation<T, L, ?> getConfiguredBaseOperation(String namespace, OperationContext context, Class<T> apiTypeClass, Class<L> apiListTypeClass) {
+  private static <T> boolean isKeyOfType(String key, Class<T> apiTypeClass) {
+    String plural = CustomResource.getPlural(apiTypeClass);
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java:350]: Calling `CustomResource#getPlural` from here is a clear symptom that this method should be moved to `HasMetadata`.

Proper steps would be to move the method to HasMetadata class, keep the method in CustomResource class and call the newer in HasMetadata while marking it as deprecated with instructions to call the new one.

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java (hunk 1 above)
Comment (metacosm) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java:65]: The use case would be similar to the case above with `CustomResource.getPlural`: when we get a class that's not accurately typed but we still want to retrieve possibly existing information. Now, the question is: should the classes always be properly typed or not?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java (hunk 3 above)
Comment (metacosm) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java:350]: Agreed, initially I thought that it only made sense on `CustomResource` but it makes also sense on `HasMetadata` when I think more about it.

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java (hunk 1 above)
Comment (manusa) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java:65]: Well, at the very least if we are dealing with some Kubernetes resource, they should implement the HasMetadata interface, most of the client expectations depend on that. (so not the same case as `Class<? extends CustomResource>`)

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java (hunk 3 above)
Comment (metacosm) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java]: Done.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Comment by rohanKanojia** [comment]
some test seem to be failing:
```
[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    UtilsTest.testGetPluralFromKind:225->lambda$testGetPluralFromKind$0:226 expected: <componentstatuses> but was: <componentstatus>
Error:    SerializationTest.unmarshalCRDWithSchema:72 expected:<...ue
    storage: true[]> but was:<...ue
    storage: true[
plural: "customresourcedefinitions"
singular: "customresourcedefinition"]>
[INFO] 
Error:  Tests run: 352, Failures: 2, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Fabric8 :: Kubernetes :: Project 5.3-SNAPSHOT:
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Utils.java [hunk 4]
```diff
@@ -345,32 +346,8 @@ public static String toUrlEncoded(String str) {
...
-      case "PodSecurityPolicy":
-      case "ServiceEntry": // an Istio resource. Needed as getPluralFromKind is currently not configurable #2489
-        // Delete last character
-        pluralBuffer.deleteCharAt(pluralBuffer.length() - 1);
-        pluralBuffer.append("ies");
-        break;
-      case "Endpoints":
-        break;
-      default:
-        pluralBuffer.append("s");
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Utils.java:371]: Related to the new test failure,

should we preserver the code block and set default to:
```java
    default:
        return Pluralize.toPlural(kind.toLowerCase(Locale.ROOT));
```

@rohanKanojia @metacosm ???

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/UtilsTest.java [hunk 5]
```diff
@@ -170,7 +171,7 @@ void interpolateStringWithParametersTest() {
   @Test
   void testGetPluralFromKind() {
     // Given
-    Map<String, Class> pluralToKubernetesResourceMap = new HashMap<>();
+    SortedMap<String, Class> pluralToKubernetesResourceMap = new TreeMap<>();
```
Comment (manusa) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/UtilsTest.java:174]: This test is strongly calling for a refactor to make it aparameterized test.

We'll need to revisit this in a future PR.

**Review by manusa — APPROVED** [review]

**Comment by rohanKanojia** [comment]
looks good to me :heavy_check_mark:

**Pull request merged**